monai/data/meta_tensor.py

```diff
@@ -0,0 +1,257 @@
+    """
+    global _TRACK_TRANSFORMS
+    _TRACK_TRANSFORMS = val
+
+
+def get_track_meta() -> bool:
+    """
+    Get track meta data boolean.
+    """
+    global _TRACK_META
+    return _TRACK_META
+
+
+def get_track_transforms() -> bool:
+    """
+    Get track transform boolean.
+    """
+    global _TRACK_TRANSFORMS
+    return _TRACK_TRANSFORMS
+
+
+class MetaTensor(torch.Tensor):
+    """
+    Class that extends upon `torch.Tensor`, adding support for meta data.
+
+    We store the affine as its own element, so that this can be updated by
+    transforms. All other meta data that we don't plan on touching until we
+    need to save the image to file lives in `meta`.
+
+    Behavior should be the same as `torch.Tensor` aside from the extended
+    functionality.
+
+    Copying metadata:
+        * For `c = a + b`, then the meta data will be copied from the first
+        instance of `MetaTensor`.
+    """
+
+    AffineType = torch.Tensor
+    MetaType = TypeVar("MetaType", dict, list[dict])
+    TransformsType = TypeVar("TransformsType", list[Callable], list[list[Callable]])
+
+    _affine: AffineType
+    _meta: MetaType
+    _transforms: TransformsType
+
+    @staticmethod
+    def get_default_affine() -> AffineType:
+        return torch.eye(4)
+
+    @staticmethod
+    def get_default_meta() -> dict:
+        return {}
+
+    @staticmethod
+    def get_default_transforms() -> list[Callable]:
+        return []
+
+    def set_initial_val(self, attribute: str, input_arg: Any, input_tensor: Any, default_fn: Callable):
+        """
+        Set the initial value. Try to use input argument, but if this is None
+        and there is a MetaTensor input, then copy that. Failing both these two,
+        use a default value.
+        """
+        if input_arg is None:
+            input_arg = getattr(input_tensor, attribute, None)
+        if input_arg is None:
+            input_arg = default_fn()
+        setattr(self, attribute, input_arg)
+
+    @staticmethod
+    def __new__(
+        cls,
+        x,
+        affine: AffineType | None = None,
+        meta: MetaType | None = None,
+        transforms: TransformsType | None = None,
+        *args,
+        **kwargs,
+    ) -> MetaTensor:
+        return torch.as_tensor(x, *args, **kwargs).as_subclass(cls)  # type: ignore
+
+    def __init__(
+        self,
+        x,
+        affine: AffineType | None = None,
+        meta: MetaType | None = None,
+        transforms: TransformsType | None = None,
+    ) -> None:
+        """If `affine` is given, use it. Else, if `affine` exists in the input tensor, use it. Else, use
+        the default value. The same is true for `meta` and `transforms`."""
+        self.set_initial_val("affine", affine, x, self.get_default_affine)
+        self.set_initial_val("meta", meta, x, self.get_default_meta)
+        self.set_initial_val("transforms", transforms, x, self.get_default_transforms)
+
+    @staticmethod
+    def get_meta_tensors(args):
```

Comment:
May need to support nested meta info.
Similar to bundle access: https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/config_parser.py#L118.

Thanks.